### DISCUSSION TITLE
cycle between 2 different widgets on mouse click.

### DISCUSSION THREAD

**Discussion by Sydiepus** [qtile/qtile#3545] [Q&A] [answered]
Hello i asked a similar question [here](https://github.com/qtile/qtile/discussions/2857) however this time i want to switch between to 2 different widgets the Wlan and the Net widgets.
I tried the following but the Net is not updating at all, sometimes it goes up to 1KB randomly then drops but most of the time its stuck on 0.00B on both up and down i have another Net widget next to it that updates just fine.
```py
def net_wid(toggled=False):
    global tog
    if toggled :
        tog = not tog
    if not tog :
        return widget.Wlan(interface="wlan0",format="  {essid} {percent:2.0%}").poll()
    else :
        return widget.Net(interface="wlan0",format="{down} ↓↑{up}").poll()
tog = False
...
 widget.GenPollText(func=net_wid, update_interval=1,mouse_callbacks={"Button1": lambda : net_wid(True)}),
```
Edit: the widget was not working not sure how i messed but it should work now.

**Comment by elParaguayo** [comment, marked as answer]
Interesting approach! 

The issue is that, every time you call `net_wid` you're creating a new instance of those widgets. That's no go good for `Net` as it reports the differences between each call to `poll`. You always get 0 because it's the first poll.

Try creating the widget outside of your function and then calling `poll` on that instance in `net_wid`.

**Reply by Sydiepus** [reply]
I tried the following :
```py
def net_wid(toggled=False):
    global tog
    Netwid = qtile.widgets_map["net"]
    if toggled :
        tog = not tog
    if not tog :
        return widget.Wlan(interface="wlan0",format="  {essid} {percent:2.0%}").poll()
    else :
        return Netwid.poll()
```
However this leave me with 2 net widgets and the readings are inaccurate.
Is there a proper way to create a widget that don't updates nor gets displayed in the bar ?

**Reply by Sydiepus** [reply]
I tried increasing the update interval but the readings are in Bytes even though they should be way higher.

**Reply by elParaguayo** [reply]
> However this leave me with 2 net widgets and the readings are inaccurate. 

Why 2? Don't put it in the bar. I meant something like this:

``` python
Netwid = widget.Net(interface="wlan0",format="{down} ↓↑{up}")

def net_wid(toggled=False):
    global tog
    if toggled :
        tog = not tog
    if not tog :
        return widget.Wlan(interface="wlan0",format="  {essid} {percent:2.0%}").poll()
    else :
        return Netwid.poll()
```

> Is there a proper way to create a widget that don't updates nor gets displayed in the bar ?

Not really. Widgets are designed specifically to go in the bar. You're trying to use widgets in a way that is not intended so I'm not entirely surprised that you're getting odd results.

Also, bear in mind that your Net widget will not be polling in the background so, when you switch to it for the first time, you're going to get an odd result.

**Reply by Sydiepus** [reply]
That's weird I'm sure I tried this and ended up having an error about tuples not having a poll attribute. 
It's working now, thank you.
Answer for anyone in the future :
```py
Netwid = widget.Net(interface="wlan0",format="{down} ↓↑{up}")

def net_wid(toggled=False):
    global tog
    if toggled :
        tog = not tog
    if not tog :
        return widget.Wlan(interface="wlan0",format="  {essid} {percent:2.0%}").poll()
    else :
        return Netwid.poll()
tog = False
...
widget.GenPollText(func=net_wid, update_interval=1,mouse_callbacks={"Button1": lambda : net_wid(True)}),
```

**Reply by m-col** [reply]
> Also, bear in mind that your Net widget will not be polling in the background so, when you switch to it for the first time, you're going to get an odd result.

For this reason I am a bit confused as to why this works. Shouldn't the widget always be static and never update itself, if it wasn't in the bar when the bar is setup?

**Reply by elParaguayo** [reply]
No. The GenPollText widget is calling the function regularly so the the Net widget is being pulled as a result of that, not its own timers.

**Reply by m-col** [reply]
ah, cool. I guess that's in part lucky - many widgets would fail on account on not having `self.qtile` or `self.bar` set, presumably

**Reply by elParaguayo** [reply]
Exactly.
